src/cares_wrap.h

```diff
@@ -0,0 +1,639 @@
+      void* arg,
+      int status,
+      int timeouts,
+      struct hostent* host) {
+    QueryWrap<Traits>* wrap = FromCallbackPointer(arg);
+    if (wrap == nullptr) return;
+
+    struct hostent* host_copy = nullptr;
+    if (status == ARES_SUCCESS) {
+      host_copy = node::Malloc<hostent>(1);
+      cares_wrap_hostent_cpy(host_copy, host);
+    }
+
+    wrap->response_data_ = std::make_unique<ResponseData>();
+    ResponseData* data = wrap->response_data_.get();
+    data->status = status;
+    data->host.reset(host_copy);
+    data->is_host = true;
+
+    wrap->QueueResponseCallback(status);
+  }
+
+  void QueueResponseCallback(int status) {
+    BaseObjectPtr<QueryWrap<Traits>> strong_ref{this};
+    env()->SetImmediate([this, strong_ref](Environment*) {
+      AfterResponse();
+
+      // Delete once strong_ref goes out of scope.
+      Detach();
+    });
+
+    channel_->set_query_last_ok(status != ARES_ECONNREFUSED);
+    channel_->ModifyActivityQueryCount(-1);
+  }
+
+  void CallOnComplete(
+      v8::Local<v8::Value> answer,
+      v8::Local<v8::Value> extra = v8::Local<v8::Value>()) {
+    v8::HandleScope handle_scope(env()->isolate());
+    v8::Context::Scope context_scope(env()->context());
+    v8::Local<v8::Value> argv[] = {
+      v8::Integer::New(env()->isolate(), 0),
+      answer,
+      extra
+    };
+    const int argc = arraysize(argv) - extra.IsEmpty();
+    TRACE_EVENT_NESTABLE_ASYNC_END0(
+        TRACING_CATEGORY_NODE2(dns, native), trace_name_, this);
+
+    MakeCallback(env()->oncomplete_string(), argc, argv);
+  }
+
+  // Subclasses should implement the appropriate Parse method.
+  void Parse(unsigned char* buf, int len) {
+    Traits::Parse(this, buf, len);
+  }
+
+  void Parse(struct hostent* host) {
+    Traits::Parse(this, host);
+  }
+
+  void MemoryInfo(MemoryTracker* tracker) const override {
+    tracker->TrackField("channel", channel_);
+    if (response_data_) {
+      tracker->TrackFieldWithSize("response", response_data_->buf.size);
+    }
+  }
+
+  SET_MEMORY_INFO_NAME(QueryWrap)
+  SET_SELF_SIZE(QueryWrap<Traits>)
+
+ private:
+  BaseObjectPtr<ChannelWrap> channel_;
+
+  std::unique_ptr<ResponseData> response_data_;
+  const char* trace_name_;
+  // Pointer to pointer to 'this' that can be reset from the destructor,
+  // in order to let Callback() know that 'this' no longer exists.
+  QueryWrap<Traits>** callback_ptr_ = nullptr;
+};
+
+template <typename Impl>
+struct QueryTraits final {
+  static constexpr const char* name = Impl::name;
+
+  static int Send(
+      QueryWrap<QueryTraits<Impl>>* wrap,
+      const char* name) {
+    wrap->AresQuery(name, Impl::dns_class, Impl::type);
+    return 0;
+  }
+
+  static void Parse(
+      QueryWrap<QueryTraits<Impl>>* wrap,
+      unsigned char* buf,
+      int len) {
+    Impl::Parse(wrap, buf, len);
+  }
+
+  static void Parse(
+      QueryWrap<QueryTraits<Impl>>* wrap,
+      struct hostent* host) {
+    UNREACHABLE();
+  }
+};
```

Comment:
This class seems like an unnecessary indirection? It doesn’t seem to provide anything that we couldn’t put into `Impl` or `QueryWrap` itself

Reply:
Possibly. `ReverseWrap` is the only variant that uses the `hostent` variant of `Parse` and that variant is the only reason why this is added. I can take a pass at refactoring this out. I hadn't already because I was trying to minimize the changes to `QueryWrap`.

Reply:
Ok, the intermediate struct has been removed.